Source/WebCore/page/Navigation.cpp

```diff
@@ -120,26 +120,38 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
+
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(frame()->frameID() == currentItem->frameID());
+
     // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
     if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
         bool shouldProcessPreviousNavigationEntries = [&]() {
             if (!previousNavigation->m_entries.size())
                 return false;
-            if (navigationType == NavigationNavigationType::Traverse)
-                return false;
             if (!frame()->document()->protectedSecurityOrigin()->isSameOriginAs(previousWindow->document()->protectedSecurityOrigin()))
                 return false;
             return true;
         }();
         if (shouldProcessPreviousNavigationEntries) {
             for (auto& entry : previousNavigation->m_entries)
                 m_entries.append(NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), entry.get()));
-            m_currentEntryIndex = previousNavigation->m_currentEntryIndex;
-            if (navigationType == NavigationNavigationType::Reload) {
-                updateForActivation(currentItem.get(), navigationType);
-                return;
+
+            Ref previousEntry = m_entries[*previousNavigation->m_currentEntryIndex];
+
+            if (navigationType == NavigationNavigationType::Push) {
+                m_entries.resize(*previousNavigation->m_currentEntryIndex + 1); // Prune forward entries.
```

Comment:
I did not explicitly address pruning in https://github.com/WebKit/WebKit/pull/35304, maybe my PR is failing some cases because of that.